core/corehttp/gateway_handler_unixfs_dir.go

```diff
@@ -133,20 +139,92 @@ func (i *gatewayHandler) serveDirectory(ctx context.Context, w http.ResponseWrit
 			ShortHash: shortHash(hash),
 		}
 		dirListing = append(dirListing, di)
+		entryRequests = append(entryRequests, &entryRequest{
+			entryIndex: len(dirListing) - 1,
+			cid:       link.Cid,
+		})
+	}
+
+	// Optimization 2:
+	// Fetch in parallel entry's metadata up to FastDirIndexThreshold.
+	// Code inspired from go-merkledag.Walk().
+	// FIXME: Review number and (maybe) export it.
+	const defaultConcurrentFetch = 32
+	fetchersCtx, cancel := context.WithCancel(ctx)
+	defer cancel()
+	feed := make(chan *entryRequest)
+	go func() {
+		for _, r := range entryRequests {
+			select {
+			case feed <- r:
+			case <-fetchersCtx.Done():
+				return
+			}
+		}
+	}()
+	type entryInfo struct {
+		entryIndex int
+		size string
+	}
+	out := make(chan entryInfo)
+	errChan := make(chan error)
+
+	for c := 0; c < defaultConcurrentFetch; c++ {
+		go func() {
```

Comment:
Do we really need that many goroutines ?
Why not use `GetMany` ? which accepts a list of CIDs and return a channel.

Reply:
> Do we really need that many goroutines ?

The number is up for discussion, I don't think 32 is not _that_ many though in terms of performance impact if that's your concern (the number was extracted from the DAG walk for example).

> Why not use GetMany ? 

I'm not sure I want to subscribe to 10K requests in BS and then cancel the context with the remaining un-served ones once we reach the `FastDirIndexThreshold` threshold, but I'm definitely not happy with the complexity I'm adding here. If lidel greenlights this approach I'll look for some inbetween like batching some number of requests in each goroutine with `GetMany`.

Reply:
I was expecting something really simple, like `GetMany` them 32 by 32 sequentially. I think this approach would even be simpler than what you wrote.

I don't think it's worth fixing in this PR, someone else will file a new PR if that approach is found to

Reply:
A sequential approach is what we're changing away from. If you request 32 blocks, 31 are delivered and one is missing you're stuck just the same than if you request one at a time like the previous implementation.

Reply:
> If you request 32 blocks, 31 are delivered and one is missing you're stuck

AFAIK, That already what we do for `ipfs pin add` because of the scheduling costs gains.

Resolving, this only need to be looked at if that show up in someone's profile.

Reply:
Being discussed in https://github.com/ipfs/go-ipfs/pull/8888#issuecomment-1101591394 for lidel to resolve.